test/e2e_node/pod_termination_state_test.go

```diff
@@ -0,0 +1,182 @@
+
+	"github.com/onsi/ginkgo/v2"
+)
+
+// Pod Termination State Test will cover both pod termination with graceful
+// shutdown and oomkilled cases, in both cases, it will verify
+// 1. pod is in termating status after receiving sig term
+// 2. cgroups associated with the pod are deleted
+// 3. pod is no longer existed in API server after termination
+var _ = SIGDescribe("Pod Termination State [NodeConformance]", func() {
+	f := framework.NewDefaultFramework("pod-termination-state-test")
+	f.NamespacePodSecurityEnforceLevel = admissionapi.LevelPrivileged
+
+	// Normal Pod Cases
+	normalContainerName := "normal-container"
+	podSpec := getNormalPod("normal-pod", normalContainerName)
+	runNormalTest(f, podSpec, normalContainerName)
+
+	// Evicted Pod Cases (oomkill)
+	containerName := "oomkill-target-container"
+	oomPodSpec := getOOMTargetPod("oomkill-target-pod", containerName)
+	runEvictedTest(f, oomPodSpec, containerName)
+})
+
+func runNormalTest(f *framework.Framework, podSpec *v1.Pod, containerName string) {
+	ginkgo.Context("", func() {
+		var podClient *e2epod.PodClient
+		ginkgo.BeforeEach(func() {
+			ginkgo.By("setting up the pod to be used in the test")
+			podClient = e2epod.NewPodClient(f)
+			podClient.Create(context.TODO(), podSpec)
+		})
+
+		ginkgo.It("The pod should be running and deleted with grace period", func() {
+
+			// We need to wait for the pod to be running, otherwise the deletion
+			// may be carried out immediately rather than gracefully.
+			framework.ExpectNoError(e2epod.WaitForPodNameRunningInNamespace(context.TODO(), f.ClientSet, podSpec.Name, f.Namespace.Name))
+
+			// save the running pod
+			pod, err := podClient.Get(context.TODO(), podSpec.Name, metav1.GetOptions{})
+			framework.ExpectNoError(err, "failed to GET scheduled pod")
+
+			// save the UID for cgroup verify
+			podUID := string(pod.UID)
+
+			// delete pod gracefully
+			ginkgo.By("deleting the pod gracefully")
+			var lastPod v1.Pod
+			var statusCode int
+			err = f.ClientSet.CoreV1().RESTClient().Delete().AbsPath("/api/v1/namespaces", pod.Namespace, "pods", pod.Name).Param("gracePeriodSeconds", "30").Do(context.TODO()).StatusCode(&statusCode).Into(&lastPod)
+			framework.ExpectNoError(err, "failed to use http client to send delete")
+			framework.ExpectEqual(statusCode, http.StatusOK, "failed to delete gracefully by client request")
+
+			// verifying the kubelet observed the termination notice
+			// check pod status is from running to terminating status, after a while, pod is deleted in kubelet
+			err = e2epod.WaitForPodTerminatingInNamespaceTimeout(context.TODO(), f.ClientSet, podSpec.Name, f.Namespace.Name, 30)
+			framework.ExpectNoError(err)
+
+			// Sleep enough time to ensure the pod is terminated
+			time.Sleep(30 * time.Second)
+
+			// resources are cleaned up and phase is set correctly cgroup files
+			// see same way in pods_container_manager_test.go
+			new_pod := makePodToVerifyCgroupRemoved("pod" + podUID)
```

Comment:
how is this different then the exiting pods_container_manager_test? I believe it also creates a pod, deletes it gracefully, waits for the pod to be succeeded, and verifies the pod cgroups are deleted.

Reply:
The exiting pods_container_manager_test is more focus on cgroup creation and deletion during the pod lifecycle. 

For this test , it is more focus on termination state like checking the pod enter the terminating states after graceful shut down and the after the pod deletion, the cgroup and pod are deleted completely.